### DISCUSSION TITLE
Attention, dynamic shapes and JIT

### DISCUSSION THREAD

**Discussion by dfdx** [jax-ml/jax#18790] [Q&A] [answered]
Probably, a trivial question, but I can't the right keywords to find an answer. 

On one hand, `jax.jit` [requires all arrays to have static shape](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#dynamic-shapes). 

On another hand, attention in transformers operates on sequences of difference length. Of course, it's possible to pad all sequences to the maximum length, but if, for example, my maximum is 4096 tokens and real sequence is 30 tokens, then it's a huge waste of computational resources. 

How do people solve this dilemma?

**Comment by PhilipVinc** [comment]
Incremental padding at sizes that are either power of 2, or multiples of something else.

**Reply by dfdx** [reply]
Thank you for the reply! Let me clarify the question and add some context. I'm trying to JIT-compile my [Llama implementation in JAX](https://github.com/dfdx/gendo/blob/llama/gendo/llama/model.py), but JAX compiler complaints about dynamic shapes or arguments. If you want to know the details, it mostly complains about `freqs_cis` argument - an intermediate array from which we take a slice to compute rotary embeddings, and the shape of this slice [depends](https://github.com/dfdx/gendo/blob/llama/gendo/llama/model.py#L479) on the input sequence length. 

JAX doesn't like slices (arrays) whose shape it cannot infer statically. The shape of the intermediate array (`freqs_cis`) depends on the shape of the input sequence, but input sequence naturally varies. Padding the sequence to power of 2 won't help since inputs will still be of different lengths (e.g. 32 and 64 tokens) and JIT just won't let me compile the function. The documentation just states "don't use dynamic shapes", but in Transformers it's unavoidable. But still people somehow overcome this limitation. 

Sorry for the clunky explanation - as I admitted at the beginning, I cannot find the right keywords to explain my problem concisely. I will be happy to provide any further clarification.

**Reply by doctor-phil** [reply]
@PhilipVinc could you explain how incremental padding could be done without changing the size of the array?

**Reply by jakevdp** [reply]
If you're curious how people implement attention in JAX, you could start by looking at implementations such as this one in flax: https://github.com/google/flax/blob/829e004c6d5646f24bd048023a4407e3b4d5aa98/flax/experimental/nnx/examples/07_transformer.py#L181

**Reply by doctor-phil** [reply]
@jakevdp Thank you, I'll take a look at this and it might be helpful.

My question is not about attention, necessarily. I am just wondering if there is a systematic way to solve problems that require dynamic array sizes without padding arrays to the maximum length and doing a lot of extra calculations.

**Reply by jakevdp** [reply, marked as answer]
In general, the strategy for handling dynamic array sizes in JAX is, depending on the situation, to either dispatch a new JIT-compiled operation for each size, or to pad all entries to a maximum size. In practice, the extra calculations done in the padding strategy may not be as important as you fear, especially if you're running on an accelerator like GPU or TPU.

**Reply by dfdx** [reply]
Thank you!
